### ISSUE TITLE
[feature request] pretty evaluation overlay

### ISSUE THREAD

**Issue by bo-tato** [joaotavora/sly#567]
Emacs lisp with [eros.el](https://github.com/xiongtx/eros) and CIDER for clojure both show evaluation results in a pretty box next to the expression, which is easier to look at then looking down at the messages and back. This would be a nice addition for sly. [This comment](https://www.reddit.com/r/emacs/comments/bi4xk1/evaluation_overlays_in_slime_for_common_lisp/) shows how to do it with slime, I changed it slightly and it works with sly:
```
(defun sly-eval-sexp-overlay ()
  (interactive)
  (eros--eval-overlay
   (let ((result (s-trim (sly-eval `(slynk:pprint-eval ,(sly-sexp-at-point))))))
     (cond
      ;; number or string, eval it
      ((-contains-p '(?0 ?1 ?2 ?3 ?4 ?5 ?6 ?7 ?8 ?9 ?\")
                    (string-to-char (substring result 0 1)))
       (eval (car (read-from-string result))))

      ;; eg #\\Newline
      ((s-starts-with-p "#\\" result) result)

      ;; cf http://www.lispworks.com/documentation/HyperSpec/Body/02_dh.htm
      ;; eg #p"arst" => "#p arst"
      ((s-starts-with-p "#" result)
       (format "%s %s"
               (substring result 0 2)
               (substring result 3 (- (length result) 1))))

      ;; maybe remove NIL from the end
      ((s-ends-with-p "\nNIL" result) result)

      ;; list
      ((and (s-starts-with-p "(" result)
            (s-ends-with-p ")" result))
       (eval (car (read-from-string (format "(quote %s)" result)))))

      ;; spaces, give up
      ((s-contains-p " " result) result)

      ;; symbol
      (t (eval (car (read-from-string (format "(quote %s)" result)))))
      ))
   (point)))
```
I don't really know any emacs lisp yet but this seems a simple enough task for me to learn, if you're interested in integrating it in sly by default I can try and make a PR request for it. Maybe the right place to add it would be to in `sly-display-eval-result` check if `eros.el` is present and if so use that rather than or in addition to the standard `message`?

**Comment by bo-tato** [comment]
this doesn't print somethings properly and is also insecure if you are developing a program that prints untrusted data to standard-out (it is passing output to elisp eval). I'm now using the simpler:
```
(defun sly-eval-sexp-overlay ()
  (interactive)
  (let ((result (->> `(slynk:pprint-eval ,(sly-sexp-at-point))
                     sly-eval
                     s-trim
                     ; separate by ", " rather than newline for multi-value results
                     (s-replace "\n" ", "))))
    (eros--make-result-overlay result
      :where (point)
      :duration eros-eval-result-duration)))
```
depends on https://github.com/joaotavora/sly/pull/590 to work correctly with multi-value results

**Comment by fosskers** [comment]
I spent some time today putting together [`sly-overlay`](https://git.sr.ht/~fosskers/sly-overlay) based on EROS, and it works, but there are issues calling `sly-eval-defun` since internally it eventually calls `sly-rex` and processes the results via callbacks, giving no apparent way to just get the result value back as-is from the function to then apply an overlay (it is expecting the callback to just print the result somehow). So, all I see in the buffer as an overlay is `=> t` no matter what I execute.

**Comment by joaotavora** [comment]
? Looking forward to your non-abusive version

**Comment by fosskers** [comment]
Given the limitations of Emacs Lisp in that it has no shared memory primitives, I suppose using exceptions like this might be the only way to pass back the results eh, assuming that `sly-rex` isn't blocking. 

`sly-eval-defun` and `sly-eval` seem to have quite different execution paths, with only the latter using the exception trick. Can you elaborate on the motivation for the difference? I suppose using `sly-eval` instead might get me what I want for the time being.

Once I get `sly-overlay` working and cleaned up, I'll publish it as a contrib library, unless that's functionality you'd like in the main library (re: Clojure's CIDER)?

**Comment by bo-tato** [comment]
for reference that snippet I pasted above is what I've been using for my limited CL development since then. It mostly works but there's things I want to polish on it that I haven't gotten around to:
1. `slynk:pprint-eval` is synchronous, not callback with result like sly's normal eval commands, so if eval something slow then need to unhang emacs with `C-G`
2. stdout gets shown in the eros overlay, not the repl. I'd rather make stdout go to the repl as normal and just show the result in the overlay

**Comment by fosskers** [comment]
@bo-tato My `sly-overlay` port borrows everything that EROS can do, so if I can complete it then that might solve your needs. I'll try again soon with `sly-eval` internally.

**Comment by bo-tato** [comment]
I think clojure just doesn't have multiple values, you can just return a list or vector to be a bit less efficient for the same use cases? All my eval-defun, eval-sexp etc commands are calling that `sly-eval-sexp-overlay`, so those are probably the same issues you'll run into with using `sly-eval` but I think they're all fixable. Thanks for your work

**Comment by fosskers** [comment]
Got it! I used your example as a base and got it working. I'll polish this up and publish it.

**Comment by joaotavora** [comment]
> Given the limitations of Emacs Lisp in that it has no shared memory primitives, I suppose using exceptions like this might be the only way to pass back the results eh, assuming that sly-rex isn't blocking.

Just wanted to say that this has nothing to do with shared memory (also what languages have "shared mamory primitives", afaik it's always a library  feature).  Anyway this is just a non-local return into another frame of the same call stack.  Emacs is single threaded so if you want synchronous operation without busy waiting (and ensuing deadlock) this is how it's done.  The technique is inherited directly from SLIME and is much cleaned up in `jsonrpc.el`.  See `jsonrpc-request`.

But there is also `sly-eval-async` (and `jsonrpc-async-request` btw) if you don't want the sync behaviour.  I don't know exactly the problem you're trying to solve.  But `sly-eval-async` is usually the way to go if you want the user to continue using Emacs while something is ongoing.

**Comment by fosskers** [comment]
> also what languages have "shared mamory primitives", afaik it's always a library feature

Off the top of my head, Rust and Haskell do.

>  Anyway this is just a non-local return into another frame of the same call stack

Thanks for the elaboration.

>  I don't know exactly the problem you're trying to solve.

It was that using `sly-eval-defun` was asynchronous, but the overlay logic borrowed from EROS wanted some result synchronously in order to overlay it into the buffer. Got it figured out though, here's the meat of it: https://git.sr.ht/~fosskers/sly-overlay/tree/master/item/sly-overlay.el#L208-224

**Comment by joaotavora** [comment]
[> > also what languages have "shared mamory primitives", afaik it's always a library feature
> Off the top of my head, Rust and Haskell do.

Both a library feature.  Anyway, I don't know how you'd solve anything with shared memory here: SLY connects through the network, so the address space you'd want to share could be on the other side of the ocean.

> here's the meat of it

If this is an interactive on-demand thing?  Couldn't tell from the description.  "Eros" seems to be automatic and @bo-tato's version plugs into that.   I'd appreciate some kind of video or animation showing what Eros and `sly-overlay.el` does.

@bo-tato mentions the sync/async issue.  I'd use an async version so you can keep working with Emacs while a slow thing is executing over on the Slynk side.   The overlay could still be added if the buffer is live and in the right conditions.  Or not, maybe you don't mind blocking the user.  Also `SLYNK:EVAL-AND-GRAB-OUTPUT` is better if you want to go through `*ECHO-NUMBER-ALIST*` and other protocols, discard or keep output.

**Comment by fosskers** [comment]
Here's what an overlay looks like in the buffer when calling `sly-overlay-eval-defun`:

![sly-overlay](https://github.com/joaotavora/sly/assets/229679/805135e0-9b9a-4ab8-bc58-5cce218cb9c4)

In this case I'm just using `sly-eval` internally and blocking the client. I believe that CIDER and EROS also do the same.

**Comment by joaotavora** [comment]
EROS seems to be on-demand and interactive too, indeed.  Still maybe a prettier version of this would put some `<computing...>` overlay right away then come back to update it later on, idk.    OTOH if the expressions errors I suppose it's nice to get the debugger as usual.  Or maybe not, maybe the overlay could come back as `<error, click here if you care>`.  Anyway, good luck.

I don't think this feature is mature enough to make it into SLY proper, maybe a contrib.

**Comment by fosskers** [comment]
We do still get the debugger popping up as expected when there's an error, which is nice.

> I don't think this feature is mature enough to make it into SLY proper, maybe a contrib.

Agreed, I'll publish it separately. Thanks for your input and thanks for Sly.

**Comment by bo-tato** [comment]
thanks I tried it out, it's nicer and has better overlay placement than my little hack, but it still seems to have most of the other same issues? it's synchronous and hangs until the eval finishes, stdout no longer shows up in the sly repl, it doesn't distinguish when the result is multiple values vs multiple lines, and display of large results doesn't work well with the overlay. For example in elisp with eros overlay when you do `(s-lines (f-read "/etc/passwd"))`, it sees it's a big result and displays the normal non-overlay way. imho the normal `sly-eval-defun` display is nicer for large results like `(uiop:read-file-lines "/etc/passwd")` than seeing a truncated result in the overlay so it should display the normal way if the result is too big for the overlay. Also I'd need a generic `sly-overlay-sexp` rather than just eval defun as I have a handful of other custom eval functions rather than just eval top level form.

When I have the time to work on it I'll much prefer to contribute to your package rather than my separate hack. I haven't used sourcehut before but am happy to try alternatives to github, does it also have an issue tracker and PR or patch submission system there, or how does contribution work?

Closing this issue as like you say it's suited as a sly-contrib rather than merging into sly proper for now. Is there some awesome-sly listing or something that lists all the different sly-contrib's people have made? I haven't found one, if it doesn't exist I think it'd be worth to make.
